[otbn/doc] Secure Wipe documentation update. (#7288)
lowRISC/opentitan: hw/ip/otbn/doc/_index.md

Refer to the [Data Integrity Protection]({{<relref "#design-details-data-integrity-protection">}}) section for details of how the data integrity protections are implemented.

## Secure Wipe

OTBN provides a mechanism to securely wipe all state it stores, including the instruction memory.

> This mechanism is a superset of other secure wipe mechanisms:

Reviewer: I'm not sure this is quite the right phrasing. Maybe just "This wiping is split into three parts"?

- Data Memory Secure Wipe
- Instruction Memory Secure Wipe

> - Internal State Secure Wipe

Reviewer: Could these items link to the relevant sections? Also, are the capitals really necessary? It sounds a bit shouty to me.

> The secure wipe mechanism can be triggered by host software through the bus interface, and is automatically initiated in situations where OTBN has determined that the state it stores may have been compromised.

Reviewer: Please add a link to the relevant register for the bus interface.

> The secure wipe mechanism is triggered when executing the [`ECALL`]({{< relref "isa#ecall" >}}) instruction.

Reviewer: I think this is just the internal state secure wipe, right?

> The secure wipe mechanism can be triggered by host software through the bus interface using register {{< regref "SEC_WIPE">}}, and is automatically initiated in situations where OTBN has determined that the state it stores may have been compromised.
> 
> The internal state secure wipe mechanism is triggered when executing the [`ECALL`]({{< relref "isa#ecall" >}}) instruction.

Reviewer 2: This is effectively all one paragraph explaining the trigger mechanisms. Maybe make that clearer with text like this?

```suggestion
A secure wipe is performed automatically in certain situations, or can be requested manually by the host software.
A secure wipe of all state is performed when a fatal error is detected. 
(TODO: Specify lifecycle state dependency.)
A secure wipe of the internal state only is performed whenever an OTBN operation finishes.
Finally, host software can manually trigger a secure wipe operation by writing to the {{< regref "SEC_WIPE">}} register.
```

(Leave the TODO in there for now, it's something we need to figure out later.)

Reviewer 2: Thinking further: we certainly want to do a internal secure wipe on a recoverable error (since that's identical to just terminating an OTBN operation early). We might even want to do a full secure wipe in that case. The list of errors which cause a recoverable error is at https://docs.opentitan.org/hw/ip/otbn/doc/index.html#Reg_err_bits and they could be both a programmer error, or a side-effect of a fault injection. 

The answer probably brings us back to the debuggability vs. security discussion.

@vrozic can you update the document to always include internal wipe for recoverable alerts, and leave a TODO for the potential full secure wipe? I could imagine the outcome being "full secure wipe if in PROD mode, otherwise internal wipe only."

Refer to the [Data Integrity Protection]({{<relref "#design-details-data-integrity-protection">}}) section for details of how the data integrity protections are implemented.

## Secure Wipe

OTBN provides a mechanism to securely wipe all state it stores, including the instruction memory.

The full secure wipe mechanism is split into three parts:
- [Data memory secure wipe]({{<relref "#design-details-secure-wipe-dmem">}})
- [Instruction memory secure wipe]({{<relref "#design-details-secure-wipe-imem">}})
- [Internal state secure wipe]({{<relref "#design-details-secure-wipe-internal">}})

A secure wipe is performed automatically in certain situations, or can be requested manually by the host software.
The full secure wipe is automatically initiated as a local reaction to a fatal error.
(TODO: Specify reaction to incoming lifecycle escalation event, and a potential lifecycle state dependency, e. g. only wipe in PROD mode.)

> A secure wipe of the internal state is performed whenever an OTBN operation is completed and after a recoverable error.

Reviewer 2: nit:

```suggestion
A secure wipe of only the internal state is performed whenever an OTBN operation is completed and after a recoverable error.
```

Finally, host software can manually trigger the full or partial secure wipe operations by writing to the {{< regref "SEC_WIPE">}} register. 
During the secure wipe operation the {{< regref "STATUS.busy">}} flag is set to 1, indicating that the OTBN is busy.

> Once the operation is completed, a {{< regref "INTR_STATE.done" >}} interrupt is set and {{< regref "STATUS.busy">}} is cleared. 

Reviewer 2: The section this text appears in is the high-level description of security features. I would argue that the busy/interrupt behavior is better placed in the design details section further down.

And a nit:
```suggestion
Once the operation is completed, a {{< regref "INTR_STATE.done" >}} interrupt is raised and {{< regref "STATUS.busy">}} is cleared. 
```

On that topic: It is already implicitly defined, but maybe add one explicit sentence that OTBN being busy during the secure wipe operation, and writes to SEC_WIPE being ignored while OTBN is busy, means effectively that host software will get a single done interrupt for a secure wipe operation, independent of how many SEC_WIPE bits the software wrote.

> Refer to the [Secure wipe]({{<relref "#design-details-secure-wipe">}}) section for implementation details.

Reviewer 2: ```suggestion
Refer to the [Secure Wipe]({{<relref "#design-details-secure-wipe">}}) section for implementation details.
```

(To match the section title below.)

The scrambling is reversed on a read.

The ephemeral memory scrambling key and the nonce are provided by the [OTP block]({{<relref "/hw/ip/otp_ctrl/doc" >}}).

> They are set once when OTBN block is reset, and changed whenever a [secure wipe]({{<relref "#design-details-secure-wipe">}}) of the data memory is performed.

Reviewer: This link is to the wrong section: it should point at the DMEM secure wipe, I think.

The scrambling is reversed on a read.

The ephemeral memory scrambling key and the nonce are provided by the [OTP block]({{<relref "/hw/ip/otp_ctrl/doc" >}}).

> They are set once when OTBN block is reset, and changed whenever a [secure wipe]({{<relref "#design-details-secure-wipe">}}) of the instruction memory is performed.

Reviewer: Again, wrong section: should be the IMEM secure wipe, I think.

The Integrity Protection Code is checked on every memory read, even though the code remains attached to the data.
A further check must be performed when the data is consumed.
Detected integrity violations in the data memory raise a fatal `imem_error`.

### Secure Wipe {#design-details-secure-wipe}

> 

Reviewer 2: Can you add a one-line intro sentence to secure wipe here?

In the following situations OTBN itself initiates a full secure wipe:
* The lifecycle controller asks for it through its escalation signal.
* A fatal alert is issued. 
  In this case, a full secure wipe is performed as local action.

> Software can trigger a full state wipe by writing 3’b111 to {{< regref "SEC_WIPE">}}  (i.e. set all individual state wipe bits).

Reviewer 2: Three comments:
* Can you put this content into a separate section, e.g. "Triggering secure wipe"
* Please avoid non-ASCII characters.
* Define "full secure wipe" somewhere.

> #### Data Memory (DMEM) Secure Wipe

Reviewer: Could we rename this and the IMEM one to something like "DMEM key rotation"? I think it would make the other text a bit clearer, avoiding repeating various types of "wipe", which gets a bit confusing.

Reviewer 2: Hm, I'm not sure how much we want to mix implementation details (key rotation) with what it achieves (get rid of data).

Reviewer: Hmm, maybe. The problem is that I got quite confused when reading through this patch and I'm supposed to know what's going on already! I think it would really help if the names were more distinct.

The wiping is performed by securely replacing the memory scrambling key, making all data stored in the memory unusable. 
The key replacement is a two-step process:

* Overwrite the 128b key of the memory scrambling primitive with randomness from URND.

>   This action is deterministic in time and takes a single cycle.

Reviewer: The "deterministic in time" phrasing is a bit clunky and is redundant, given that this takes a single cycle. Maybe just "This takes a single cycle."?

Author: corrected in the latest version

* Request new scrambling parameters from OTP. 
  The request takes multiple cycles to complete.

> Host software can initiate a data secure wipe by writing 1 to the SEC_WIPE.DMEM register field.

Reviewer: I think you mean "data memory secure wipe" here? But, as I said above, I think this would be much clearer if we called it something like "DMEM key rotation".

#### Instruction Memory (IMEM) Secure Wipe

The wiping is performed by securely replacing the memory scrambling key, making all instructions stored in the memory unusable. 
The key replacement is a two-step process:

* Overwrite the 128b key of the memory scrambling primitive with randomness from URND.

>   This action is deterministic in time and takes a single cycle.

Reviewer: Again, maybe just "This takes a single cycle."?

Author: corrected

* Request new scrambling parameters from OTP.
  The request takes multiple cycles to complete.

> Host software can initiate a data secure wipe by writing 1 to the SEC_WIPE.IMEM register field.

Reviewer: Wrong name here.

Author: corrected

#### Internal State Secure Wipe {#design-details-secure-wipe-internal}

OTBN provides a mechanism to securely wipe all internal state, excluding the instruction and data memories.

The following state is wiped:
* Register files: GPR and WDR
* The accumulator register (also accessible through the ACC WSR)

> * Software-writable special-purpose registers: the FG0, FG1, FLAGS, and MOD[0-7] registers

Reviewer 2: Many of these registers are actually mappings of each other and point to the same data. So instead of using the register names, maybe something like this would work better:

* Flags (accessible through the FG0, FG1, and FLAGS CSRs)
* The modulus (accessible through the MOD0 to MOD7 CSRs and the MOD WSR)

> The wiping procedure is a two step process:

Reviewer 2: ```suggestion
The wiping procedure is a two-step process:
```

The scrambling is reversed on a read.

The ephemeral memory scrambling key and the nonce are provided by the [OTP block]({{<relref "/hw/ip/otp_ctrl/doc" >}}).
They are set once when OTBN block is reset, and changed whenever a [secure wipe]({{<relref "#design-details-secure-wipe-imem">}}) of the instruction memory is performed.

The Integrity Protection Code is checked on every memory read, even though the code remains attached to the data.
A further check must be performed when the data is consumed.
Detected integrity violations in the data memory raise a fatal `imem_error`.

### Secure Wipe {#design-details-secure-wipe}

> The entire OTBN internal state including the instruction and data memories, can be securely deleted on demand from a host software.  

Reviewer 2: We use the term "internal state" in this context as separate category next to instruction and data memories.

```suggestion
The entire OTBN state, including the contents of the instruction and data memories, can be securely deleted on demand from a host software.  
```

In addition, full or partial secure wipe is triggered automatically by the OTBN in certain situations. 

#### Triggering Secure Wipe

In the following situations OTBN itself initiates a full secure wipe:
* The lifecycle controller asks for it through its escalation signal.
* A fatal alert is issued. 
  In this case, a full secure wipe is performed as a local action.

> Internal state secure wipe is automatically triggered on completion of the OTBN operation and after a recoverable error alert.

Reviewer 2: Two small things:
* We are not super consistent, but generally tried to go with "error" as the condition that is observed, and "alert" being the reaction to an error.
* For recoverable errors, we still "complete" the operation, just not successfully.

```suggestion
The internal state secure wipe is automatically triggered when an OTBN operation completes, either successfully, or unsuccessfully due to a recoverable error.
```

Reviewer 2: @vrozic did you intentionally not update this part, or did you just overlook it?

> Software can trigger a full state wipe by writing 3'b111 to {{< regref "SEC_WIPE">}}  (i.e. set all individual state wipe bits).

Reviewer 2: ```suggestion
Host software can trigger a full state wipe by writing 3'b111 to {{< regref "SEC_WIPE">}}  (i.e. set all individual state wipe bits).
```

#### Data Memory (DMEM) Secure Wipe {#design-details-secure-wipe-dmem}

The wiping is performed by securely replacing the memory scrambling key, making all data stored in the memory unusable. 
The key replacement is a two-step process:

* Overwrite the 128b key of the memory scrambling primitive with randomness from URND.
  This action takes a single cycle.
* Request new scrambling parameters from OTP. 
  The request takes multiple cycles to complete.

> Host software can initiate a data memory secure wipe by writing 1 to the {{< regref "SEC_WIPE.dmem">}} register field.

Reviewer 2: Add a reference here to a section describing how completion of the operation is signaled.

#### Instruction Memory (IMEM) Secure Wipe {#design-details-secure-wipe-imem}

The wiping is performed by securely replacing the memory scrambling key, making all instructions stored in the memory unusable. 
The key replacement is a two-step process:

* Overwrite the 128b key of the memory scrambling primitive with randomness from URND.
  This action takes a single cycle.
* Request new scrambling parameters from OTP.
  The request takes multiple cycles to complete.

> Host software can initiate an instruction memory secure wipe by writing 1 to the {{< regref "SEC_WIPE.imem">}} register field.

Reviewer 2: As above

#### Internal State Secure Wipe {#design-details-secure-wipe-internal}

OTBN provides a mechanism to securely wipe all internal state, excluding the instruction and data memories.

The following state is wiped:
* Register files: GPRs and WDRs
* The accumulator register (also accessible through the ACC WSR)
* Flags (accessible through the FG0, FG1, and FLAGS CSRs)
* The modulus (accessible through the MOD0 to MOD7 CSRs and the MOD WSR)

The wiping procedure is a two-step process:
* Overwrite the state with randomness from URND.
* Overwrite the state with zeros.

Loop and call stack pointers are reset.

> Host software can initiate an internal state secure wipe by writing 1 to the {{< regref "SEC_WIPE.internal">}} register field.

Reviewer 2: And as above again.

They are set once when OTBN block is reset, and changed whenever a [secure wipe]({{<relref "#design-details-secure-wipe-imem">}}) of the instruction memory is performed.

The Integrity Protection Code is checked on every memory read, even though the code remains attached to the data.
A further check must be performed when the data is consumed.
Detected integrity violations in the data memory raise a fatal `imem_error`.

### Secure Wipe {#design-details-secure-wipe}

The entire OTBN state, including the contents of instruction and data memories, can be securely deleted on demand from a host software.
In addition, full or partial secure wipe is triggered automatically by the OTBN in certain situations.

#### Triggering Secure Wipe

In the following situations OTBN itself initiates a full secure wipe:
* The lifecycle controller asks for it through its escalation signal.
* A fatal alert is issued.
  In this case, a full secure wipe is performed as a local action.

The internal state secure wipe is automatically triggered when an OTBN operation completes, either successfully, or unsuccessfully due to a recoverable error.

> Host software can trigger a full state wipe by writing 3'b111 to {{< regref "SEC_WIPE">}}  (i.e. set all individual state wipe bits).

Reviewer: Nit: Could you surround the 3'b111 with back-ticks? Otherwise, the apostrophe character will get rendered like normal punctuation, which might look a bit odd.

#### Completion of Secure Wipe {#design-details-secure-wipe-completion}

During the secure wipe operation the {{< regref "STATUS.busy">}} flag is set to 1, indicating that the OTBN is busy.
Once the operation is completed, a {{< regref "INTR_STATE.done" >}} interrupt is raised and {{< regref "STATUS.busy">}} is cleared.
This effectively means that the host software will get a single done interrupt for a secure wipe operation, independent of how many SEC_WIPE bits the software wrote.

#### Data Memory (DMEM) Secure Wipe {#design-details-secure-wipe-dmem}

The wiping is performed by securely replacing the memory scrambling key, making all data stored in the memory unusable.
The key replacement is a two-step process:

* Overwrite the 128b key of the memory scrambling primitive with randomness from URND.
  This action takes a single cycle.
* Request new scrambling parameters from OTP.
  The request takes multiple cycles to complete.

Host software can initiate a data memory secure wipe by writing 1 to the {{< regref "SEC_WIPE.dmem">}} register field.

> The [completion]({{<relref "#design-details-secure-wipe-completion">}}) is signalled by raising a {{< regref "INTR_STATE.done" >}} interrupt.

Reviewer: Nit: raising a**n** (because INTR_STATE starts with a vowel). Similarly for the next two sections.
